PDF context menu should be shown asynchronously

aprotyas opened the pull request:
#### 3babd344c6c08255c3ce2db7459b57d99209ae43
<pre>
PDF context menu should be shown asynchronously
<a href="https://bugs.webkit.org/show_bug.cgi?id=268973">https://bugs.webkit.org/show_bug.cgi?id=268973</a>
<a href="https://rdar.apple.com/122525514">rdar://122525514</a>

Reviewed by Simon Fraser.

WebPageProxy::ShowPDFContextMenu is a sync IPC call that blocks the WP.
Among other things, this manifests itself in repaint requests queued
right before the IPC call to be fulfilled only _after_ a context menu
action has been taken.

There is no reason for this synchronicity, so this patch async-ifies
WebPageProxy::ShowPDFContextMenu, which we call with sendWithAsyncReply
to learn about the context menu item selected by the user.

We also make some drive-by fixes such as avoiding blindly casting
integral values from the UIP into ContextMenuItemTag and moving some
function definitions around so we can annotate function groups with
`#pragma mark`.

* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleContextMenuEvent):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::handleMouseEnterEvent):
(WebKit::UnifiedPDFPlugin::handleMouseLeaveEvent):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::contextMenuItemTagFromDisplayMode const):
(WebKit::UnifiedPDFPlugin::displayModeFromContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: <a href="https://commits.webkit.org/274320@main">https://commits.webkit.org/274320@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c3bdc6957f556075aba0f8d8ec751bd3c4f96fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17637 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41237 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14983 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39279 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14863 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12900 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42513 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35160 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36917 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8675 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1092,7 +1092,7 @@ static bool getEventTypeFromWebEvent(const WebEvent& event, NSEventType& eventTy
 
     NSUserInterfaceLayoutDirection uiLayoutDirection = webPage->userInterfaceLayoutDirection() == UserInterfaceLayoutDirection::LTR ? NSUserInterfaceLayoutDirectionLeftToRight : NSUserInterfaceLayoutDirectionRightToLeft;
-    NSMenu *nsMenu = [m_pdfLayerController menuForEvent:nsEventForWebMouseEvent(event) withUserInterfaceLayoutDirection:uiLayoutDirection];
+    RetainPtr<NSMenu> nsMenu = [m_pdfLayerController menuForEvent:nsEventForWebMouseEvent(event) withUserInterfaceLayoutDirection:uiLayoutDirection];
```
You should be able to just say `RetainPtr nsMenu = `

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/9c3bdc6957f556075aba0f8d8ec751bd3c4f96fc

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1213,12 +1248,21 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    return isKnownContextMenuItemTag ? static_cast<ContextMenuItemTag>(tagValue) : ContextMenuItemTag::Unknown;
+}
+
 PDFContextMenu UnifiedPDFPlugin::createContextMenu(const IntPoint& contextMenuPoint) const
```
I had to go look to see what type `PDFContextMenu` was, to check the memory safety. It's not a menu! It's more like `PDFContextMenuInfo` or `PDFContextMenuData`

aprotyas replied:
Yes `PDFContextMenu` and `NSMenu` are quite different object types. Retaining the `NSMenu` should be enough for memory safety!

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1213,12 +1248,21 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
 
+auto UnifiedPDFPlugin::toContextMenuItemTag(int tagValue) const -> ContextMenuItemTag
+{
+    static constexpr std::array regularContextMenuItemTags { ContextMenuItemTag::OpenWithPreview, ContextMenuItemTag::SinglePage, ContextMenuItemTag::SinglePageContinuous, ContextMenuItemTag::TwoPages, ContextMenuItemTag::TwoPagesContinuous, ContextMenuItemTag::ZoomIn, ContextMenuItemTag::ZoomOut, ContextMenuItemTag::ActualSize };
```
I would unwrap this to have one item per line.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/9c3bdc6957f556075aba0f8d8ec751bd3c4f96fc

aprotyas committed:
PDF context menu should be shown asynchronously
https://bugs.webkit.org/show_bug.cgi?id=268973
rdar://122525514

Reviewed by Simon Fraser.

WebPageProxy::ShowPDFContextMenu is a sync IPC call that blocks the WP.
Among other things, this manifests itself in repaint requests queued
right before the IPC call to be fulfilled only _after_ a context menu
action has been taken.

There is no reason for this synchronicity, so this patch async-ifies
WebPageProxy::ShowPDFContextMenu, which we call with sendWithAsyncReply
to learn about the context menu item selected by the user.

We also make some drive-by fixes such as avoiding blindly casting
integral values from the UIP into ContextMenuItemTag and moving some
function definitions around so we can annotate function groups with
`#pragma mark`.

* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleContextMenuEvent):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::handleMouseEnterEvent):
(WebKit::UnifiedPDFPlugin::handleMouseLeaveEvent):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::contextMenuItemTagFromDisplayMode const):
(WebKit::UnifiedPDFPlugin::displayModeFromContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: https://commits.webkit.org/274320@main
